[WIP] Add GSI space charge benchmark lattice

cemitch99 opened the pull request:
Added the lattice used for space charge benchmarks described here:
http://web-docs.gsi.de/~giuliano/research_activity/trapping_benchmarking/main.html

- [ ] The problem parameters need to be double-checked. In particular, verify the longitudinal dynamics.
- [x] Implement first for a 2.5D space charge solver? #1214
- [x] test of linear (envelope) optics w/o space charge, w/o sextupole
- [x] test of orbits (resonance behavior) w/o space charge, w/ sextupole
- [x] test of tune calculation w/o space charge, w/o sextupole
- [x] test of tune calculation w/ space charge, w/o sextupole
- [x] test of tune calculation w/ space charge, w/o sextupole using Gauss2p5d solver
- [ ] tests w/ space charge, w/ sextupole

There is much that can be useful here.  For now, data will be archived here, and we can assess later what is most useful to leave in the examples directory.  I will move the status to WIP for the time being.

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add GSI space charge benchmark lattice.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete input_kurth_10nC.in

This is not part of this pull request.

cemitch99 committed:
Add rf and modify distribution input parameters.

cemitch99 committed:
Merge branch 'development' into add_gsi_benchmark_lattice

cemitch99 commented:
This was sitting on the back shelf of the "to do" list.  Now it's time to revisit.

cemitch99 committed:
Add Python draft of this lattice.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update lattice files to use Twiss and validate.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
`setup.py`: Emscripten (#1528)

Support a cross-compiled Python include in WASM Pyodide
Emscripten builds in `setup.py`. Try to auto-derive from
environment hints and platform in the sysconfig of the Python
interpreter.

cemitch99 committed:
Separate benchmark for envelope optics comparison.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add files for nonlinear phase space benchmark.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Performance: Optimize (Chr)Quad (#1521)

* Performance: Optimize (Chr)Quad, PlasmaLens

Reduce branching in Quad, ChrQuad and ChrPlasmaLens.

* Revert: ChrPlasmaLens

Avoid computing unneeded transcendentals...

* Quad[spin]: Vectorize for DP

Only DP+nospin does not pay off to vectorize.

* (Chr)Quad: Cleaning

cemitch99 committed:
Add species to MAD-X input; use MAD-X matched dispersion values.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Move distribution to separate file; clean-up.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add draft of tune diagnostics.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Clean-up of in-situ script.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Fix Pandas sorting in in-situ tune calculation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/gsi_benchmark/run_gsilattice_in_situ_tunes.py:
```diff
@@ -0,0 +1,146 @@
+    # To plot the normalized data:
+    # plt.scatter(xn, pxn, s=5)
+    output = pnf.naff(z, turns=n, nterms=4, skipTurns=0, getFullSpectrum=True, window=1)
+    tune = output[0, 1]
```
Analysis: since we only read `output[0, 1]` we can be 3.4x faster when doing:
```suggestion
    output = pnf.naff(z, turns=n, nterms=1, skipTurns=0, getFullSpectrum=True, window=1)
    tune = output[0, 1]
```
Each extra term is another FFT + the expensive  #`frefin` refinement loop that we do not use.

ax3l replied:
Claude bug comment:
  - You never remove the DC/closed-orbit component. With an offset and nterms=1, naff returned an empty array  (0,) → output[0,1] would IndexError. After z -= z.mean() it correctly returned tune 0.317. (This is likely  why you reached for nterms=4 — it was masking the DC line. Subtract the mean and nterms=1 is both faster   and robust.)

ax3l replied:
I added a full-change suggestion below, so do _not_ apply this one-liner here please.

cemitch99 replied:
Got it.  All looks reasonable to me--I'll accept here, and then run some local tests.

cemitch99 replied:
Note:  These Claude-suggested changes do not appear to affect the output.

ax3l commented on examples/gsi_benchmark/run_gsilattice_in_situ_tunes.py:
```diff
@@ -0,0 +1,146 @@
+    z = xn - 1j * pxn
+    # To plot the normalized data:
+    # plt.scatter(xn, pxn, s=5)
+    output = pnf.naff(z, turns=n, nterms=4, skipTurns=0, getFullSpectrum=True, window=1)
```
Claude:
`turns=n` is an off-by-one. PyNAFF needs `len(data) ≥ turns+1`. When `n` is a multiple of 6 it crashes , otherwise it silently rounds turns down to a multiple of 6 and drops up to 5 turns. Use turns=`len(z)-1`.

ax3l commented on examples/gsi_benchmark/run_gsilattice_in_situ_tunes.py:
```diff
@@ -0,0 +1,146 @@
+
+# clean shutdown
+sim.finalize()
+
+df = pd.DataFrame(test_data, columns=["turn", "idcpu", "x", "px"])
+
+# Group by particle ids, without changing their relative order
+grouped_by_idcpu = df.groupby("idcpu", sort=False)
+
+# all particle ids (once)
+particle_ids = df["idcpu"].unique().tolist()
+
+# matched Twiss functions
+alpha_x = 1.29174698
+beta_x = 12.79711091
+
+n = len(turn_arr)
+# loop over one particle track at a time
+for idcpu, particle_track in grouped_by_idcpu:
+    xarr = np.array(particle_track["x"])
+    pxarr = np.array(particle_track["px"])
+    xn = xarr / np.sqrt(beta_x)
+    pxn = pxarr * np.sqrt(beta_x) + xarr * alpha_x / np.sqrt(beta_x)
+    z = xn - 1j * pxn
+    # To plot the normalized data:
+    # plt.scatter(xn, pxn, s=5)
+    output = pnf.naff(z, turns=n, nterms=4, skipTurns=0, getFullSpectrum=True, window=1)
+    tune = output[0, 1]
+    plt.scatter(xarr[0] * mm_scale, tune)
+
+plt.xlabel("x [mm]", fontsize=12)
```
Claude suggested improvements: 2 bug fixes, small perf improvement, works with pyNAFF <1.2.0, too:
```suggestion
beta_x = 12.79711091
sqrt_beta = np.sqrt(beta_x)              # hoist out of the loop

x0_mm, tunes = [], []                    # accumulate, plot once
for idcpu, track in grouped_by_idcpu:
    xarr  = track["x"].to_numpy()        # cheaper than np.array(track["x"])
    pxarr = track["px"].to_numpy()
    
    xn  = xarr / sqrt_beta
    pxn = pxarr * sqrt_beta + xarr * alpha_x / sqrt_beta
    z   = xn - 1j * pxn
    z  -= z.mean()                        # remove closed-orbit / DC line
    
    out = pnf.naff(
        z,
        turns=len(z) - 1,                 # FIX: turns+1 must be <= len(z)
        nterms=1,                         # FIX: you only use term 0  -> 3.4x faster
        getFullSpectrum=True,             # required: z is complex
        window=1,
        warnings=False,
    )
    if len(out) == 0:                     # no line (e.g. lost/clipped particle)
        continue
    x0_mm.append(xarr[0] * mm_scale)
    tunes.append(out[0, 1])

plt.scatter(x0_mm, tunes, s=5)            # single draw call (was 1 per particle)
plt.xlabel("x [mm]", fontsize=12)
```

cemitch99 committed:
Apply Claude suggestions re PyNAFF processing.

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Thin Elements: No Collective Kicks (#1533)

Do no collective kicks in the 1-slice loop when dealing
with thin elements. Kicks for those live in the drifts
around them.

cemitch99 committed:
Add in-situ tune calculation with space charge included.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete MADX Twiss data.

This data is available on the benchmark website, and can also be reproduced from the MADX input file.

ax3l committed:
Array: Fixed Seed (#1536)

Fix #1326

ax3l committed:
CI: Win MSVC w/ AVX2 (#1539)

ax3l committed:
openPMD-api: 0.17.1 (#1535)

Build in superbuild, but do not require, the lastest.

cemitch99 committed:
Add space charge w/o sextupole (using 2.5D PIC).

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Move distribution information into localized function.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Move distribution in remaining examples to local Python function.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on examples/gsi_benchmark/distribution_gsi.py:
```diff
@@ -0,0 +1,46 @@
+    # bunch_charge_C = 4.4e-9  #This works with 10K particles.  With 100K particles, it is too high.
+    # bunch_charge_C = 3.7e-9  #4 nC = 0.24, 3.5 nC = 0.26..)
+    bunch_charge_C = 4.2e-9  # Try this with 100K particles. 3.7 nC gives 0.26, increase. 4 nC gives 2.55, increase
+    beam_current_A = 1.5  # used with 2D space charge
```
```suggestion
```

cemitch99 committed:
Remove unused beam current.

cemitch99 committed:
Add version of sc test using Gauss2p5d solver.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Fix handling of user apply_long... flag.

cemitch99 committed:
Remove debug print statement.

cemitch99 committed:
Remove unused comment; modify naming of variable in Python binding.

cemitch99 committed:
Add a small test.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add analysis script.

cemitch99 committed:
Print absolute value dpt_max to avoid normalizing by 0.

cemitch99 committed:
Delete unused lines from analysis script.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Turn of z-to-t transformation.

cemitch99 committed:
Minor analysis script clean-up.

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

cemitch99 committed:
Restore computation of zi.

cemitch99 committed:
Restore computation of zi in analysis script.

cemitch99 committed:
More minor analysis script clean-up.

cemitch99 committed:
Merge branch 'development' into add_gsi_benchmark_lattice

cemitch99 committed:
Merge branch 'development' into add_gsi_benchmark_lattice

cemitch99 committed:
Add usage of monitor for tbt moments only.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Merge branch 'development' into add_gsi_benchmark_lattice

cemitch99 committed:
Add the case of space charge w sextupole.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
